Assets/LeapMotionModules/UI/Interaction/InteractionBehaviour.cs

```diff
@@ -0,0 +1,692 @@
+        holdingMovementController.MoveTo(newPosition, newRotation, this);
+
+        OnGraspedMovement(origPosition, origRotation, newPosition, newRotation, hand);
+      }
+
+      OnGraspHold(hand);
+    }
+
+    public override void GraspEnd(Hand hand) {
+      _graspCount--;
+
+      InteractionHand graspingIntHand = interactionManager.GetInteractionHand(hand);
+      _graspingIntHandsCache.Remove(graspingIntHand);
+
+      if (moveObjectWhenGrasped) {
+        GraspedPoseController.RemoveHand(graspingIntHand);
+      }
+
+      // Revert kinematic state if the grasp has ended
+      if (_graspCount == 0) {
+        Rigidbody.isKinematic = _wasKinematicBeforeGrab;
+      }
+
+      OnGraspEnd(hand);
+    }
+
+    // TODO: Implement suspend/resume callbacks and example behavior.
+    public override void GraspSuspendObject(Hand hand) {
+      throw new System.NotImplementedException();
+    }
+
+    public override void GraspResumeObject(Hand hand) {
+      throw new System.NotImplementedException();
+    }
+
+    #endregion
+
+    #region Internal
+
+    protected Transform[] _childrenArray;
+
+    protected void InitInternal() {
+      _childrenArray = GetComponentsInChildren<Transform>(true);
+
+      FixedUpdateLayer();
+    }
+
+    #region Locked Position Checking
+
+    private bool _isPositionLocked = false;
+
+    /// <summary>
+    /// Returns whether the InteractionBehaviour has its position fully locked
+    /// by its Rigidbody settings or by any attached PhysX Joints.
+    /// 
+    /// This is useful for the GraspedMovementController to determine whether
+    /// it should attempt to move the interaction object or merely rotate it.
+    /// 
+    /// If the state of the underlying Rigidbody or Joints changes what this value
+    /// should be, it will not automatically update (as an optimization) at runtime;
+    /// instead, manually call RefreshPositionLockedState(). This is because the
+    /// type-checks required are relatively expensive and mustn't occur every frame.
+    /// </summary>
+    public bool isPositionLocked { get { return _isPositionLocked; } }
+
+    /// <summary>
+    /// Call this method if the InteractionBehaviour's Rigidbody becomes or unbecomes
+    /// fully positionally locked (X, Y, Z) or if a Joint attached to the Rigidbody
+    /// no longer locks its position (e.g. by being destroyed or disabled).
+    /// </summary>
+    public void RefreshPositionLockedState() {
+      if ((_body.constraints & RigidbodyConstraints.FreezePositionX) > 0
+       && (_body.constraints & RigidbodyConstraints.FreezePositionY) > 0
+       && (_body.constraints & RigidbodyConstraints.FreezePositionZ) > 0) {
+        _isPositionLocked = true;
+        return;
+      }
+      else {
+        Joint[] joints = _body.GetComponents<Joint>();
+        foreach (var joint in joints) {
+          FixedJoint fixedJoint = joint as FixedJoint;
+          if (fixedJoint != null) {
+            _isPositionLocked = true; return;
```

Comment:
Would argue that setting it to false at the top and then setting it to true after would be slightly better.  That way if someone comes in here and makes a change and forgets to add a return, it doesn't suddenly break, it just gets slightly slower.

Also I think it would be good to put these returns on their own line.